indico/modules/events/persons/client/js/Email.jsx

```diff
@@ -0,0 +1,264 @@
+        </Translate>
+      )}
+      {open && (
+        <EmailForm
+          eventId={eventId}
+          personIds={personIds}
+          userIds={userIds}
+          roleIds={[roleId]}
+          onClose={() => setOpen(false)}
+          extraParams={extraParams}
+        />
+      )}
+    </>
+  );
+}
+
+EmailButton.propTypes = {
+  eventId: PropTypes.number.isRequired,
+  roleId: PropTypes.number,
+  personSelector: PropTypes.string,
+  userSelector: PropTypes.string,
+  triggerSelector: PropTypes.string,
+  extraParams: PropTypes.object,
+};
+
+EmailButton.defaultProps = {
+  roleId: undefined,
+  personSelector: undefined,
+  userSelector: undefined,
+  triggerSelector: undefined,
+  extraParams: undefined,
+};
+
+export function EmailForm({eventId, personIds, roleIds, userIds, onClose, extraParams}) {
+  const [preview, setPreview] = useState(null);
+  const recipientData = {personId: personIds, roleId: roleIds, userId: userIds, ...extraParams};
+  const {data, loading} = useIndicoAxios({
+    url: emailAttributesURL({event_id: eventId}),
+    params: snakifyKeys(recipientData),
+  });
+  const {
+    senders = [],
+    recipients = [],
+    subject: defaultSubject,
+    body: defaultBody,
+    placeholders = [],
+  } = data || {};
+  const count = Object.values(recipientData).reduce((acc, v) => acc + v.length, 0);
+
+  const togglePreview = async ({body, subject}) => {
+    if (!preview) {
+      const {data} = await indicoAxios.post(emailPreviewURL({event_id: eventId}), {body, subject});
+      setPreview(data);
+      return;
+    }
+    setPreview(undefined);
+  };
+
+  const onSubmit = async data => {
+    try {
+      await indicoAxios.post(emailAttributesURL({event_id: eventId}), snakifyKeys(data), {
+        params: snakifyKeys(recipientData),
+      });
+      setTimeout(() => onClose(), 5000);
+    } catch (err) {
+      return handleSubmitError(err);
+    }
+  };
+
+  const extraActions = (
+    <Form.Field styleName="preview-button">
+      <FormSpy subscription={{values: true}}>
+        {({values}) => (
+          <Button type="button" active={!!preview} onClick={() => togglePreview(values)}>
+            {!preview ? <Translate>Preview</Translate> : <Translate>Edit</Translate>}
+          </Button>
+        )}
+      </FormSpy>
+    </Form.Field>
+  );
+
+  const previewRender = preview && (
+    <>
+      <Message info>
+        <Translate as={Message.Header}>
+          This preview is only shown for a single recipient.
+        </Translate>
+        <Translate>
+          When sending the emails, each recipient will receive an email customized with their
+          personal data.
+        </Translate>
+      </Message>
+      <Form.Field required>
+        <Translate as="label">Subject</Translate>
+        {preview.subject}
+      </Form.Field>
+      <Form.Field required>
+        <Translate as="label">Email body</Translate>
+        <TextEditor
+          value={preview.body}
+          config={{showToolbar: false}}
```

Comment:
Why hide the toolbar? Isn't this an HTML email?

Reply:
I'm hiding the toolbar just for the preview, I think it looks cleaner...
Is there a better way to display the preview without using a `<TextEditor>`?

Reply:
nevermind, I missed the fact that this is for the preview